### PR TITLE
apiserver: protect registry cleanup against concurrent access

### PR DESCRIPTION
**Pull request by sttts**
The mechanism uses global state during testing. Better protect it and fail early when a race is happening.

Related to https://github.com/kubernetes/kubernetes/issues/54095

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go [hunk 1]
```diff
@@ -82,25 +84,36 @@ func StorageWithCacher(capacity int) generic.StorageDecorator {
...
 
+var cleanupLock sync.Mutex
 var cleanup []func() = nil
 
 func TrackStorageCleanup() {
+	cleanupLock.Lock()
+	defer cleanupLock.Unlock()
+
+	if cleanup != nil {
+		panic("Conflicting storage tracking")
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go]: Why not allow multiple cleanups and just only respect the first?

**Comment by deads2k** [comment]
just a question.  lgtm, pending the answer.

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go]: We use the global state and have no knowledge to which "session" the storage belongs. The whole mechanism was a quickfix with this assumption in mind. Wiring through storage shutdown properly has been done by @frobware, but introduces quite some ugliness. So @dims added this hack for the tests. We still don't see complete stability, so this let's the mechanism fail early, but with a clear message.

**Pull request merged**